### PR TITLE
Prepare deployment

### PR DESCRIPTION
**Pull request by rmeissner**
- Added audit report
- Added tasks for vesting pool
  - Show vesting information 
  - Generate deployment transaction data
  - Generate vesting creation transaction data

### CODE REVIEW

File: src/tasks/vesting_pool.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+import { calculateProxyAddress, encodeMultiSend, MetaTransaction } from "@gnosis.pm/safe-contracts";
+import { calculateVestingHash } from "../utils/hash";
+import { Vesting } from "../utils/types";
+
+const getDeployerAddress = async (hre: HardhatRuntimeEnvironment): Promise<string | undefined> => {
+    const getter = hre.config.deterministicDeployment
+    if (!getter) return "0x4e59b44847b379578588920ca78fbf26c0b4956c"
+    const chainId = await hre.getChainId()
+    if (typeof getter === "function") {
+        console.log(getter(chainId))
```
Comment (mikhailxyz) [inline src/tasks/vesting_pool.ts:18]:
```suggestion
        console.log("getDeployerAddress -> getter: ", getter(chainId))
```

**Review by mikhailxyz — APPROVED** [review]

File: src/tasks/vesting_pool.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+            ethers.utils.keccak256(deploymentCode)
+        )
+        console.log("Expected target address", targetAddress)
+        console.log("Transaction target", deployerAddress)
+        console.log("Transaction data", encodedSalt + deploymentCode.toString().slice(2))
+    });
+
+const calculateRequiredTokens = (inputs: { amount: string }[]): BigNumber => {
+    let sum = BigNumber.from(0)
+    for(const input of inputs) {
```
[the remark this answers is not in the document]
Comment (rmeissner) [reply src/tasks/vesting_pool.ts:114]: no ... it is a solidity project ... therefore only solidity files are linted :D 

Probably makes sense to add js/ts linting, but I would do that in a separate PR.

**Review by schmanu — COMMENTED** [review]
I have some questions about some implementation details / a new task to only generate and initiate the Airdrop root hash.

File: src/tasks/vesting_pool.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+        const inputs: { 
+            owner: string | undefined,
+            amount: string,
+            nonce: string,
+            curveType: string | undefined,
+            startDate: string | undefined,
+            duration: string | undefined,
+            managed: boolean | undefined,
+            expectedSafe: boolean | undefined
```
Comment (schmanu) [inline src/tasks/vesting_pool.ts:145]:
```suggestion
            expectedSafe: string | undefined
```
I think this should be a string

File: src/tasks/vesting_pool.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+import { HardhatRuntimeEnvironment } from "hardhat/types";
+import { nameToAddress } from "../utils/tokenConfig";
+import { compatHandler, multiSendCallOnlyLib, proxyFactory, readCsv, safeSingleton, writeTxBuilderJson } from "./task_utils";
+import { calculateProxyAddress, encodeMultiSend, MetaTransaction } from "@gnosis.pm/safe-contracts";
+import { calculateVestingHash } from "../utils/hash";
+import { Vesting } from "../utils/types";
+
+const getDeployerAddress = async (hre: HardhatRuntimeEnvironment): Promise<string | undefined> => {
+    const getter = hre.config.deterministicDeployment
+    if (!getter) return "0x4e59b44847b379578588920ca78fbf26c0b4956c"
```
[the remark this answers is not in the document]
Comment (rmeissner) [reply src/tasks/vesting_pool.ts:15]: It is the default that we use :)

File: src/tasks/vesting_pool.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            const vesting = {
+                account: vestingTarget,
+                curveType: input.curveType || taskArgs.defaultCurve,
+                managed: input.managed !== undefined ? input.managed : taskArgs.defaultManaged,
+                durationWeeks: input.duration || taskArgs.defaultDuration,
+                startDate: input.startDate || taskArgs.defaultStartDate,
+                amount: input.amount
+            }
+            vestings.push({ vestingHash: calculateVestingHash(vestingPool, vesting, 1), vesting})
+            txs.push({ to: vestingPool.address, data: vestingData, operation: 0, value: "0" })
```
[the remark this answers is not in the document]
Comment (rmeissner) [reply src/tasks/vesting_pool.ts:200]: The airdrop related tasks will be part of a separate PR, these tasks are only for the vesting pool.

File: src/tasks/vesting_pool.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Uxio0) [reply src/tasks/vesting_pool.ts:15]: I'm curious about this too

File: src/tasks/vesting_pool.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+const calculateRequiredTokens = (inputs: { amount: string }[]): BigNumber => {
+    let sum = BigNumber.from(0)
+    for(const input of inputs) {
+        sum = sum.add(BigNumber.from(input.amount))
+    }
+    return sum
+}
+
+task("build_add_vestings_tx", "Creates a multisend transaction to assign multiple vestings based on a CSV")
+    .addFlag("transferTokens", "Indicate whether a the multisend should include the transfer of the required tokens")
```
Comment (Uxio0) [inline src/tasks/vesting_pool.ts:121]:
```suggestion
    .addFlag("transferTokens", "Indicate whether the multisend should include the transfer of the required tokens")
```

File: src/tasks/vesting_pool.ts [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    .addFlag("createSafes", "Indicate whether a new Safe should be created for each vesting")
+    .addFlag("showVestings", "Indicate whether to display the information of the vestings that will be created")
+    .addPositionalParam("csv", "CSV file with the information of the Safes that should be created", undefined, types.inputFile)
+    .addParam("defaultOwner", "The default owner of the new vesting (or the Safe for the vesting) if no owner is specified", nameToAddress("Safe Foundation"), types.string, true)
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply src/tasks/vesting_pool.ts:125]: I don't know for sure either, maybe is just convenient when setting `createSafes=True` to get Safes with the owner set

**Review by schmanu — APPROVED** [review]

File: src/tasks/vesting_pool.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmeissner) [reply src/tasks/vesting_pool.ts:114]: https://github.com/safe-global/safe-token/issues/15

File: src/tasks/vesting_pool.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Uxio0) [reply src/tasks/vesting_pool.ts:125]: Ok! Thanks for explaining

**Pull request merged**